Run puppeteer and ffmpeg inside docker container

roosnic1 opened the pull request.

roosnic1 committed:
🔧 (gitignore): add node_modules and .env to .gitignore

🔧 (nvmrc): add .nvmrc file with Node.js version 20.16.0

✨ (index.js): use environment variables for SITE_TO_STREAM and STREAM_DEST

🔧 (package.json): add dotenv as a dev dependency and configure it in start script

roosnic1 committed:
✨ (Dockerfile): add Dockerfile to containerize the application
✨ (docker-compose.yml): add docker-compose.yml for multi-container setup
🔧 (index.js): add '--disable-extensions' argument to puppeteer launch options

devnoname120 commented on Dockerfile:
```diff
@@ -0,0 +1,53 @@
+COPY package*.json ./
+
+# Install app dependencies
+RUN npm ci
```
Anticipating Rudolf's comment 😄
```suggestion
RUN npm clean-install
```

RudolfSchreier replied:
👍

devnoname120 commented on Dockerfile:
```diff
@@ -0,0 +1,53 @@
+RUN npm ci
+
+# Install ffmpeg
+#RUN apt-get update && apt-get install -y ffmpeg libnss3 libatk1.0-0 libatk-bridge-2.0
```
```suggestion
```

devnoname120 commented on Dockerfile:
```diff
@@ -0,0 +1,53 @@
+    --no-install-recommends
+
+# Copy the rest of the application code to the app directory
+COPY . .
```
You might want to create a `.dockerignore` file and add `node_modules` to it to make sure that it doesn't interfere with the output of the `clean-install` command above.

You could for example base your `.dockerignore` on this one 🙂
https://github.com/luangjokaj/wordpressify/blob/master/.dockerignore

devnoname120 commented on Dockerfile:
```diff
@@ -0,0 +1,53 @@
+# Use the official Node.js LTS image as the base image
+FROM node:lts
```
I'd suggest to pin the LTS version of node to avoid unexpected regressions.
```suggestion
FROM node:20
```

devnoname120 commented on docker-compose.yml:
```diff
@@ -0,0 +1,12 @@
+      - STREAM_DEST=udp://127.0.0.1:1234?pkt_size=1316
+    ports:
+      - "1234:1234"
+    network_mode: host
```
Why `host` for the network?

RudolfSchreier replied:
Same question here! (There isn't a good emoji to express this sentiment on GitHub 😆)

devnoname120 commented on Dockerfile:
```diff
@@ -0,0 +1,53 @@
+EXPOSE 1234
+
+# Command to start the Node.js application
+CMD ["node", "index.js"]
```
```suggestion
CMD ["node", "start"]
```

--------------

It makes sure that it's in line with the `package.json` further down that already uses another command:

```json
    "start": "node -r dotenv/config index.js",
```

roosnic1 replied:
I don't want the `dotenv/config` in docker because it's not needed. That's why i start it directly and not with `npm start`

devnoname120 replied:
Ok I'm not sure what `dotenv/config` is exactly so I trust your good judgement 👍

roosnic1 replied:
`dotenv/config` is a npm package that loads the contents of the `.env` file into the current env vars.

devnoname120 reviewed (commented):
Some small comments but LGTM in general 👍

roosnic1 committed:
🔧 (docker): add .dockerignore to exclude unnecessary files from Docker build context
⬆️ (Dockerfile): update base image to node:lts-iron for improved stability
♻️ (Dockerfile): replace `npm ci` with `npm clean-install` for cleaner dependency installation

devnoname120 reviewed (approved):
🚢

RudolfSchreier commented on Dockerfile:
```diff
@@ -0,0 +1,53 @@
+    lsb-release \
+    wget \
+    xdg-utils \
+    --no-install-recommends
```
You covered 2/3 of the [recommendations the Docker gives about the usage of `apt-get`](https://docs.docker.com/build/building/best-practices/#apt-get).

Could you add `&& rm -rf /var/lib/apt/lists/*` at the end as well?
It should do nothing except reduce image size.

RudolfSchreier commented on index.js:
```diff
@@ -36,17 +36,13 @@ const { spawn } = require('child_process');
     '-f', 'image2pipe',
     '-r', '30',
```
You are specifying 30 fps as a target frame rate both here and on line 73 - but on line 73, it is very implicitly encoded as the result of 1000/30.

Could you define a constant and reuse it everywhere? I am sorry if this causes weird floating point issues.
This does look very easy to overlook if the next programmer just does a "CTRL-F" for "30".

roosnic1 replied:
I changed that part of the code already and will ask for a review in another PR.

RudolfSchreier commented on index.js:
```diff
@@ -36,17 +36,13 @@ const { spawn } = require('child_process');
-    //'srt://listener_ip:port?pkt_size=1316' // Replace with your SRT listener details
+    process.env.STREAM_DEST
   ]);
 
```
(This regards line 73, which GitHub won't let me select :()

1. Have you considered using `setInterval`?
2. Either way, it would be safest/cleanest to also save the result of `setTimeout` or `setInterval` in a variable which is then cleared when the function exits - otherwise I am not sure if this doesn't cause `captureFrame` to continuously execute even after the `ffmpeg` process has exited.
3. Also - I am not really that familiar with the logic of what happens if frame delays start piling up.
If this becomes an issue at any point, it might make sense to use a more specific method like [`lodash`'s debounce/throttle](https://css-tricks.com/debouncing-throttling-explained-examples/) or [`requestAnimationFrame` ](https://developer.mozilla.org/en-US/docs/Web/API/Window/requestAnimationFrame), but I leave this to your judgement.

RudolfSchreier reviewed (changes requested):
Only small suggestions and longer-term questions about stability.
Happy to get your feedback :)

roosnic1 committed:
Update capture strategy and clean Docker build

Refactored the frame capture method from using setTimeout to setInterval for improved consistency. Also, added a console log to track the interval ID. Enhanced Dockerfile to remove apt lists after installation, reducing image size.

RudolfSchreier commented:
@roosnic1 Just a short update since I am going on vacation for a longer period: My comments are non-blocking, and feel free to address them in a follow-up PR and override my review should you need to merge this ASAP 🙂.

🗽😎

swisstxtsokol commented:
@roosnic1 Is this PR still relevant?

devnoname120 commented:
@roosnic1 Following up on Tatjana's question — is this PR still relevant?

roosnic1 commented:
@devnoname120 nope. I'll close it.
